### PR TITLE
tests and handling closed events added

### PR DESCRIPTION
**Pull request by ali3-atlassian**
FEATURES COMPLETED:
- handling merged/declined/deleted events (registering webhooks with them and deleting branches when these webhooks are triggered)
- tests 
- branches are crossed out/line through branch instead of disappearing from screen

### CODE REVIEW

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java [hunk 1]
```diff
@@ -88,7 +89,7 @@ public void testAddPRWithDifferentKey() {
     }
 
     //no need to test adding to the store when cacheKey AND PR exists in store already because Bitbucket doesn't allow opening
-    //a new PR when there is already an opened PR.
+    //a new PR when there is already an opened PR. therefore never the case of duplicate pull requests + key
```
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:92]: what if I open a PR, then delete it, and then open a new one?

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java [hunk 2]
```diff
@@ -128,4 +129,64 @@ public void testHasOpenPRWithExistingKeyAndOpenPR() {
 
         assertTrue(pullRequestStore.hasOpenPullRequests(branchName, repository));
     }
+
+    @Test
+    public void testRemovePRWithNonExistingKey() {
+        testAddPRWithNewKey(); //there is an entry in pullRequestStore with different cacheKey
+        String newKey = "different-key";
+
+        ConcurrentMap<?, ?> oldPullRequests = pullRequestStore.getPullRequests();
```
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:138]: I think this (and a few other tests here) rely on returning the internal representation of the `pullrequestStore` we cannot do that. This test can just as well do ` pullRequestStore.removePullRequest(serverId, anotherBitbucketPullRequest);`
followed by an `assert` that the old PR is still present.

**Review by dkjellin — CHANGES_REQUESTED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java [hunk 3]
```diff
@@ -193,22 +204,15 @@ private void getJobs(RefChangedDetails refChangedDetails, BitbucketWebhookTrigge
                 .forEach(triggerDetails -> triggerDetails.getTrigger().trigger(requestBuilder.build()));
     }
 
-    private void triggerJob(PullRequestWebhookEvent event, RefChangedDetails refChangedDetails) {
+    private void triggerJob(PullRequestWebhookEvent event, RefChangedDetails refChangedDetails,
+                            Optional<BitbucketServerConfiguration> server) {
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java:208]: /nit `Optional<BitbucketServerConfiguration> serverConfiguration` or `maybeConfiguration`...

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java [hunk 4]
```diff
@@ -193,22 +204,15 @@ private void getJobs(RefChangedDetails refChangedDetails, BitbucketWebhookTrigge
...
-                                                                                        .stream()
-                                                                                        .filter(serverConfig -> refChangedDetails.getRepository()
-                                                                                                                                 .getSelfLink()
-                                                                                                                                 .contains(serverConfig.getBaseUrl()))
-                                                                                        .findFirst();
-            if (server.isPresent()) {
-                //need to do this for all config
-                pullRequestStore.addPullRequest(server.get().getId(), event.getPullRequest());
-            }
+            server.ifPresent(bitbucketServerConfiguration ->
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java:214]: ...and then you can shortern the variable in the lambda to `configuration` etc. Just a suggestion as I didn't immediately equate `server` with `serverConfiguration`

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookEndpoint.java [hunk 5]
```diff
@@ -44,7 +44,11 @@ public HttpResponse doTrigger(StaplerRequest request, StaplerResponse response)
             case MIRROR_SYNCHRONIZED_EVENT:
                 return processMirrorSynchronizedEvent(request);
             case PULL_REQUEST_OPENED_EVENT:
+            case PULL_REQUEST_MERGED_EVENT:
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookEndpoint.java:47]: There are a lot of places where you check each of the three events- MERGED, DECLINED and DELETED, despite the fact that I think Jenkins treats them all equivalently. Perhaps when the event is first marshalled, you could convert these to a generic `PULL_REQUEST_CLOSED`, so we don't need so many extra comparisons

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java [hunk 6]
```diff
@@ -146,16 +173,15 @@ private BitbucketWebhook process(WebhookRegisterRequest request,
...
 
-        Set<String> refAndPullEvent = new HashSet<>(Arrays.asList(REPO_REF_CHANGE.getEventId(), PULL_REQUEST_OPENED_EVENT.getEventId()));
-
         List<BitbucketWebhook> webhookWithRepoRefChangeOrPR = ownedHooks
                 .stream()
-                .filter(hook -> hook.getEvents().stream().anyMatch(refAndPullEvent::contains))
+                .filter(hook -> hook.getEvents().stream().anyMatch(refAndPREventIds::contains))
                 .collect(toList());
 
         if (ownedHooks.isEmpty() ||
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:187]: I'm finding this conditional a bit hard to read, as it requires referring back through the method a lot. One suggestion might be to do it like this:
`(ownedHooks.stream().noneMatch(hook -> hook.getEvents().contains(MIRROR_SYNCHRONIZED_EVENT)) && events.contains(MIRROR_SYNCHRONIZED_EVENT))`
You can do this for each conditional you have here, which removes the need for the three or four lists you have above-- but that comes at the cost of making the conditional more complex (it _definitely_ needs some comments).
Readabiliity suggestion though, if you prefer the declared lists stick with them.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java [hunk 7]
```diff
@@ -69,9 +69,20 @@ void process(MirrorSynchronizedWebhookEvent event) {
...
-        triggerJob(event, refChangedDetails);
+        Optional<BitbucketServerConfiguration> server = bitbucketPluginConfiguration.getValidServerList()
+                .stream()
+                .filter(serverConfig -> refChangedDetails.getRepository()
+                        .getSelfLink()
+                        .contains(serverConfig.getBaseUrl()))
+                .findFirst();
+        if (event.getEventKey().equals(BitbucketWebhookEvent.PULL_REQUEST_OPENED_EVENT.getEventId())) {
+            triggerJob(event, refChangedDetails, server);
+        } else {
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java:81]: Having the various closed events in a single enum would also mean you could make this else loop explicit

**Review by mhenschke-atl — APPROVED** [review]
Looks like you have addressed most of Daniel's comments- I only have a few more that are mostly optional and don't block merging. Great work!

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketPullRequest.java [hunk 8]
```diff
@@ -26,6 +28,26 @@ public BitbucketPullRequest(
...
+        }
+        BitbucketPullRequest that = (BitbucketPullRequest) o;
+        return id == that.id &&
+               fromRef.equals(that.fromRef) &&
+               toRef.equals(that.toRef);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, state, fromRef, toRef);
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketPullRequest.java:48]: You must remove the state here or you break the contract for `hashCode()` 
>If two objects are equal according to the equals(Object) method, then calling the hashCode method on each of the two objects must produce the same integer result. 

Easiest way out here is to just remove the state from the `haschCode`
(much more around this can be found in the most excellent book Effective Java, Item 8)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStore.java [hunk 9]
```diff
@@ -4,14 +4,20 @@
...
 public interface PullRequestStore {
 
     void addPullRequest(String serverId, BitbucketPullRequest pullRequest);
 
+    void removePullRequest(String serverId, BitbucketPullRequest pullRequest);
+
     boolean hasOpenPullRequests(String branchName, BitbucketSCMRepository repository);
 
-    ConcurrentMap<?, ?> getPullRequests();
+    BitbucketPullRequest getPullRequest(String key, String slug, String serverId, int requestId);
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStore.java]:
```suggestion
    BitbucketPullRequest getPullRequest(String key, String slug, String serverId, int pullRequestId);
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java [hunk 10]
```diff
@@ -28,6 +32,26 @@ public void addPullRequest(String serverId, BitbucketPullRequest pullRequest) {
...
+        Optional.ofNullable(pullRequests.get(cacheKey)).ifPresent(value -> {
+            if (value.contains(pullRequest)) {
+                value.remove(pullRequest);
+            }
+        }
+        );
+    }
+
+    /**
+     * figures out if this store contains a given branch (if it does, this means the branch has open pull requests)
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java:50]: I think this javadoc (and the one below) should be moved to the interface.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java [hunk 11]
```diff
@@ -40,11 +64,29 @@ public boolean hasOpenPullRequests(String branchName, BitbucketSCMRepository rep
...
+     * @param key
+     * @param slug
+     * @param serverId
+     * @param requestId
+     * @return desired pull request else null
+     */
     @Override
-    public ConcurrentMap<PullRequestStoreImpl.CacheKey, ConcurrentLinkedQueue<BitbucketPullRequest>> getPullRequests() {
-        return pullRequests;
+    public BitbucketPullRequest getPullRequest(String key, String slug, String serverId, int requestId) {
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java]: As mentioned during pairing, I think we should change this to return an `Optional<BitbucketPullReqest>` as that makes it adhere to the non-null return and forces callers to handle the case of not finding the PR they were looking for.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java [hunk 12]
```diff
@@ -40,11 +64,29 @@ public boolean hasOpenPullRequests(String branchName, BitbucketSCMRepository rep
...
+                return b;
+            }
+        }
+        return null;
     }
 
+    /**
+     * key for the store that distinguishes between pull requests within different repos/projects/servers
+     */
     static class CacheKey {
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java:90]: This class should be `private` the only usages I found was in the test file and they are unused at that, so I think we can make this private now.

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java [hunk 13]
```diff
@@ -42,53 +36,59 @@ private BitbucketPullRequest setupPR(String newKey) {
 
     @Test
     public void testAddPRWithNewKey() {
-        BitbucketPullRequest bitbucketPullRequest = setupPR(key);
+        BitbucketPullRequest bitbucketPullRequest = setupPR(key, BitbucketPullState.OPEN, 1);
         PullRequestStoreImpl.CacheKey cacheKey = new PullRequestStoreImpl.CacheKey(key, slug, serverId);
```
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:40]: This `cacheKey` (and every other ocarance of `CacheKey` in this class) is unused and should be removed.

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java [hunk 14]
```diff
@@ -128,4 +128,78 @@ public void testHasOpenPRWithExistingKeyAndOpenPR() {
...
+    public void testRemovePRWithExistingKeyButNonExistingPR() {
+        testAddPRWithNewKey(); //there is an entry in pullRequestStore with same cacheKey
+
+        PullRequestStore oldStore = pullRequestStore;
+        BitbucketPullRequest anotherBitbucketPullRequest = setupPR(key, BitbucketPullState.DELETED, 1);
+        PullRequestStoreImpl.CacheKey cacheKey = new PullRequestStoreImpl.CacheKey(key, slug, serverId);
+
+        pullRequestStore.removePullRequest(serverId, anotherBitbucketPullRequest);
+
+        assertEquals("Store has changed;", pullRequestStore, oldStore);
```
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:156]: this will always be true, as you pass in the same PR store (but different variable names, but they are by reference). 
We need to update this assertion to be a bit more specific.
how about asserting that
1. `anotherBitbcketPullRequest` cannot be found (returns `Optional.empty`
2. that the PR inserted with `testAddPRWithNewKey` is still present (I _think_ that's what this test is testing?)

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java [hunk 15]
```diff
@@ -42,53 +36,59 @@ private BitbucketPullRequest setupPR(String newKey) {
...
-        ConcurrentLinkedQueue queue = (ConcurrentLinkedQueue) pullRequestStore.getPullRequests().get(cacheKey);
-        assertEquals("Size mismatch for queue;", queue.size(), 1);
-        assertTrue(queue.contains(bitbucketPullRequest));
+        assertTrue(pullRequestStore.getPullRequest(key, slug, serverId, bitbucketPullRequest.getId()) == bitbucketPullRequest);
+
     }
 
     @Test
     public void testAddPRWithExistingKey() {
         testAddPRWithNewKey(); //there is an entry in pullRequestStore with same cacheKey
```
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:50]: we tend to not call tests from other tests. It is better that you extract out the relevant bits you want to reuse to a private method and call that instead.
(here and all other places)

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java [hunk 16]
```diff
@@ -128,4 +128,78 @@ public void testHasOpenPRWithExistingKeyAndOpenPR() {
...
+
+    @Test
+    public void testRemovePRWithDifferentState() {
+        BitbucketPullRequestRef bitbucketPullRequestRef = mock(BitbucketPullRequestRef.class);
+        BitbucketRepository bitbucketRepository = mock(BitbucketRepository.class);
+        BitbucketProject bitbucketProject = mock(BitbucketProject.class);
+        BitbucketPullRequest bitbucketPullRequest = new BitbucketPullRequest(1,
+                BitbucketPullState.OPEN, bitbucketPullRequestRef, bitbucketPullRequestRef);
+        BitbucketPullRequest deletepullRequest = new BitbucketPullRequest(1,
+                BitbucketPullState.OPEN, bitbucketPullRequestRef, bitbucketPullRequestRef);
```
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:195]: I think this state is wrong? the variable name indicates it should be `DELETED`

**Review by dkjellin — CHANGES_REQUESTED** [review]
I think this is starting to look good. It is mostly just small niggles left. 
A few of them really needs to be fixed (the `hasCode` one and making the `CacheKey` private springs to mind) others would be good to get fixed. 

I think some comments in the method explaining what is going on for BitbucketWebhookHandler.process (the method javadoc is really good! we just need some explanation along the way, because it is complex and hard to follow)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java [hunk 17]
```diff
@@ -108,15 +110,14 @@ private void deleteWebhooks(List<BitbucketWebhook> webhooks) {
...
     private Collection<BitbucketWebhookEvent> getEvents(WebhookRegisterRequest request) {
-        // TODO: logging for when some events are not supported
         Collection<BitbucketWebhookEvent> supportedEvents = new HashSet<>();
         if (request.isMirror()) {
             try {
                 BitbucketWebhookSupportedEvents events = serverCapabilities.getWebhookSupportedEvents();
-                Set<String> hooks = events.getApplicationWebHooks();
-                if (hooks.contains(MIRROR_SYNCHRONIZED_EVENT.getEventId())) {
+                Set<String> hooks = events.getApplicationWebHooks(); //set intersection, cast list to set set.unison
+                if (!hooks.stream().filter(MIRROR_SYNCHRONIZED_EVENT.getEventIds()::contains).collect(Collectors.toSet()).isEmpty()) {
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:118]: maybe we could do 
```
if (!Collections.disjoint(hooks, MIRROR_SYNCHRONIZED_EVENT.getEventIds())) {
```
instead? does that read better? I think it is clearer what you're trying to do? Maybe other reviewers can comment and see what they think is clearer?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java [hunk 18]
```diff
@@ -64,21 +58,16 @@ public boolean hasOpenPullRequests(String branchName, BitbucketSCMRepository rep
...
-     * @return desired pull request else null
-     */
     @Override
     public Optional<BitbucketPullRequest> getPullRequest(String key, String slug, String serverId, int pullRequestId) {
         PullRequestStoreImpl.CacheKey cacheKey =
                 new PullRequestStoreImpl.CacheKey(key, slug, serverId);
-        for (BitbucketPullRequest b: pullRequests.get(cacheKey)) {
-            if (b.getId() == pullRequestId) {
-                return Optional.of(b);
+        if (!pullRequests.containsKey(cacheKey)) {
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java:65]: There is a check-then-act race condition here. 
I think what you need to do is essentially
```java
 ConcurrentLinkedQueue<BitbucketPullRequest> prs = pullRequests.get(cacheKey) ;
        if(prs== null || prs.isEmpty()) {
            return  Optional.empty();
        }
        return prs.stream().filter(pr -> pr.getId() == pullRequestId).findFirst();
```

**Review by dkjellin — CHANGES_REQUESTED** [review]
I think this is really close now! It is only the race condition in the PR store left to fix, and even that is only used in tests so I guess we could merge with it, but I'm going ask you to fix it before merging so we don't accidentally start to use it and have bugs later.
Looks good!

**Review by dkjellin — APPROVED** [review]
Looks good. I think there are two quite minor things to fix up. I'd be happy for you to merge this, and fix those two in a separate (smaller PR) if you prefer. Or on this one whichever you prefer most.
